Documentation/tls/rotate-tls.md

```diff
@@ -1,929 +1,338 @@
 
-```
-cfssl gencert -initca etcd-ca-csr.json | cfssljson -bare etcd-ca
-```
+* Introducing the new CA to the certificate bundle of the etcd members and API server.
+* Updating the serving and client certs.
 
-This will generate several files in the current directory.
+If `ETCD_CA_CERT` was supplied to the generation script above, it will have generated a new etcd CA with associated certificates.
 
 ```
-etcd-ca-key.pem   # new private key
-etcd-ca.csr       # CSR
-etcd-ca.pem       # new certificate
-```
+generated/etcd/
+├── ca_bundle.pem
+├── patches
+│   ├── etcd-ca.patch
+│   └── etcd-client-cert.patch
+└── tls
+    ├── ca.crt
+    ├── ca.key
+    ├── ca.txt
+    ├── client.crt
+    ├── client.key
+    ├── client.txt
+    ├── peer.crt
+    ├── peer.key
+    ├── peer.txt
+    ├── server.crt
+    ├── server.key
+    └── server.txt
 
-Verify the information for the new CA certificate.
-
-```
-cfssl certinfo -cert etcd-ca.pem
+2 directories, 15 files
 ```
 
-Use the CA certificate and key to self-sign new certificates for the
-etcd cluster components.
-
-
-### Peer
-
-The peer certificate is used for etcd-to-etcd member communication.
-
-Create a Certificate Signing Request (CSR) for the etcd peers. The following is the minimum required for the CSR.
+First, introduce the new etcd CA bundle and restart the API server. This bundle includes both the new and old CA certificates.
 
 ```
-cat > etcd-peer-csr.json <<EOF
-{
-  "CN": "etcd",
-  "key": {
-    "algo": "rsa",
-    "size": 2048
-  },
-  "names": [
-    {
-      "O": "etcd"
-    }
-  ]
-}
-EOF
+kubectl --kubeconfig generated/auth/kubeconfig \
+    patch -f generated/etcd/patches/etcd-ca.patch \
+    -p "$( cat generated/etcd/patches/etcd-ca.patch )"
+kubectl --kubeconfig generated/auth/kubeconfig \
+    delete pods -n kube-system -l k8s-app=kube-apiserver
 ```
 
-Set a variable that contains the list of DNS names for each member of the etcd
-cluster. These names must be updated to match the DNS names used in the cluster.
+Etcd members also require using a bastion host for access. The same instructions as above can be used to identify the IP of a control plane instance.
 
 ```
-export ETCD_DNS_NAMES=demo-etcd-0.coreos.com,demo-etcd-1.coreos.com,demo-etcd-2.coreos.com
-```
-
-Set a variable that contains the Kubernetes service IP addresses.
+BASTION=$( kubectl --kubeconfig generated/auth/kubeconfig get nodes -o=jsonpath='{.items[*].status.addresses[?(@.type=="ExternalIP")].address}' | awk '{ print $1 }' )
+echo "Bastion is $BASTION"
 
-```
-export SERVICE_ADDRESSES=10.3.0.15,10.3.0.20
+PROXY_COMMAND="ssh -o StrictHostKeyChecking=no -q -x core@${BASTION} -W %h:22"
 ```
 
-Generate the new certificate and private key using the CFSSL utility.
+The addresses of a cluster's etcd instances be found by inspecting the API server's `--etcd-servers` flag.
```

Comment:
Inspect the API server's `--etcd-servers` flag to find the address of a cluster's etcd instances.